src/index/updater/rune_updater.rs

```diff
@@ -0,0 +1,185 @@
+use {
+  super::*,
+  crate::runes::{varint, Edict, Runestone},
+};
+
+struct Allocation {
+  balance: u128,
+  decimals: u128,
+  id: u128,
+  rune: Rune,
+}
+
+pub(super) struct RuneUpdater<'a, 'db, 'tx> {
+  id_to_entry: &'a mut Table<'db, 'tx, u64, RuneEntryValue>,
+  rune_to_id: &'a mut Table<'db, 'tx, u128, u64>,
+  outpoint_to_balances: &'a mut Table<'db, 'tx, &'static OutPointValue, &'static [u8]>,
+}
+
+impl<'a, 'db, 'tx> RuneUpdater<'a, 'db, 'tx> {
+  pub(super) fn new(
+    outpoint_to_balances: &'a mut Table<'db, 'tx, &'static OutPointValue, &'static [u8]>,
+    id_to_entry: &'a mut Table<'db, 'tx, u64, RuneEntryValue>,
+    rune_to_id: &'a mut Table<'db, 'tx, u128, u64>,
+  ) -> Self {
+    Self {
+      id_to_entry,
+      outpoint_to_balances,
+      rune_to_id,
+    }
+  }
+
+  pub(super) fn index_runes(
+    &mut self,
+    height: u64,
+    index: usize,
+    tx: &Transaction,
+    txid: Txid,
+  ) -> Result<()> {
+    let runestone = Runestone::from_transaction(tx);
+
+    // A mapping of rune ID to un-allocated balance of that rune
+    let mut unallocated: HashMap<u128, u128> = HashMap::new();
+
+    // Increment unallocated runes with the runes in this transaction's inputs
+    for input in &tx.input {
+      if let Some(guard) = self
+        .outpoint_to_balances
+        .remove(&input.previous_output.store())?
+      {
+        let buffer = guard.value();
+        let mut i = 0;
+        while i < buffer.len() {
+          let (id, len) = varint::decode(&buffer[i..])?;
+          i += len;
+          let (balance, len) = varint::decode(&buffer[i..])?;
+          i += len;
+          *unallocated.entry(id).or_default() += balance;
+        }
+      }
+    }
+
+    // A vector of allocated transaction output rune balances
+    let mut allocated: Vec<HashMap<u128, u128>> = vec![HashMap::new(); tx.output.len()];
+
+    if let Some(runestone) = runestone {
+      // Determine if this runestone conains a valid issuance
+      let mut allocation = match runestone.etching {
+        Some(etching) => {
+          // If the issuance symbol is already taken, the issuance is ignored
```

Comment:
I think unique symbols is a bad idea. A couple scenarios where this can go awry: 

- bug (or new feature) in the indexer means that previously-unrecognized issuances are now valid for new indexes
- new token issuances can be front-run and nullified either by bots watching the mempool and outbidding them, or by miners directly

Reply:
I'm really of two minds of this. You're right about the two issues you mention, along with the fact that if a project has some token name that they want, they have to come up with a new one if it's already taken.

However, unique symbols is pretty cool. Having short, human readable names you can look things up by is nice, and having a symbol, if there are duplicates, seems pretty worthless to m.

Also, duplicate symbols make fake token scams easier. There have been a bunch of scams where someone has been sent, for example, a fake USDT token, and treated it like real tether, and sent eth in return.

Reply:
Can we have unique symbols, but also make symbols not a requirement? Meaning that if a rune declares a symbol already in use or just doesn't declare a symbol by using empty string, if can only be referenced by its ID (which should be a number so as to have no possibility of colliding with symbol space)?

Reply:
> Also, duplicate symbols make fake token scams easier. There have been a bunch of scams where someone has been sent, for example, a fake USDT token, and treated it like real tether, and sent eth in return.

maybe the better solution here is to make it easier for an issuer to share an output/pubkey/etc in a format that’s consumed by wallets. If Tether can vend a json file that says “here are my issuances”, and wallets can consume that, then wallets can say “this is tether”. That blob of json could be served up on a well known url path, or via a dns record or something else. Just spitballing. 

We have the same issue with inscriptions: if someone issues a collection, anyone can make an inscription with the same artwork. So we need some offchain declaration from the artist (heres a link to the inscription) or a parent-child collection (but you need some way to get the authenticity of the parent)

Reply:
Unique symbols will make people rush in to squat anything remotely good, just like with BRC-20. It will diminish the chance of real USDT/USDC ever being issued on Runes imo

I think frauds around duplicate symbols is something that can be resolved on a higher layer. There could be public lists of notable token IDs shared by companies/DEXes/indexers, and they can show a little verified checkmark or something in the UI